### DISCUSSION TITLE
What is the state of streaming compilation support in wasmer?

### DISCUSSION THREAD

**Discussion by VoxWave** [wasmerio/wasmer#1494] [Q&A]
From my understanding wasmer doesn't currently have an interface streaming compilation on the user level. The only compilation function seems to be one that takes an array of bytes in which means that the whole webassembly module has to be in memory already.

Internally however it seems that there's some level of streaming compilation going on judging by the internal structures. The singlepass backend also seems to be something suited for streaming compilation. Is there something particular in the way of providing an interface for the user where they can provide the module as a stream of bytes or something similiar?

The application we are building has similarities to the browser in that the modules can be downloaded from a server during runtime so having streaming compilation would be useful to us. Because we are researchers, we are particularily interested in the current state and possible issues so that we can include that information in our evaluations as well as take a shot at implementing (parts of) the feature ourselves.

**Comment by syrusakbary** [comment]
Supporting streaming compilation should be easy to add once wasmparser adds support for it.

In the wasmparser side of things support is not really trivial, but they already have a PR moving towards it:
https://github.com/bytecodealliance/wasm-tools/pull/22
